ring/ingester_lifecycle.go

```diff
@@ -113,24 +136,32 @@ func (r *IngesterRegistration) pickTokens() []uint32 {
 			ringDesc = in.(*Desc)
 		}
 
-		takenTokens := []uint32{}
+		var takenTokens, myTokens []uint32
 		for _, token := range ringDesc.Tokens {
 			takenTokens = append(takenTokens, token.Token)
+
 			if token.Ingester == r.id {
-				tokens = append(tokens, token.Token)
+				myTokens = append(myTokens, token.Token)
 			}
 		}
-		if len(tokens) < r.numTokens {
-			newTokens := generateTokens(r.numTokens-len(tokens), takenTokens)
-			tokens = append(tokens, newTokens...)
+
+		if len(myTokens) > 0 {
+			log.Infof("%d tokens already exist for this ingester!", len(myTokens))
 		}
+
+		newTokens := generateTokens(r.numTokens-len(myTokens), takenTokens)
+		ringDesc.addIngester(r.id, r.hostname, r.grpcHostname, newTokens, r.state)
+
+		tokens := append(myTokens, newTokens...)
 		sort.Sort(sortableUint32(tokens))
-		ringDesc.addIngester(r.id, r.hostname, r.grpcHostname, tokens, r.state)
+
 		return ringDesc, true, nil
 	}
 	if err := r.consul.CAS(consulKey, descFactory, pickTokens); err != nil {
 		log.Fatalf("Failed to pick tokens in consul: %v", err)
+		return nil
```

Comment:
Interesting, I had the opposite reaction because it makes it look like the function is returning without an error and the program runs on. But either way, it doesn't matter much.